File: fastlane/build_info.rb
# BuildInfo management

class BuildInfo
  attr_accessor :platform
  attr_accessor :build_id
  attr_accessor :version
  attr_accessor :internal_version
  attr_accessor :short_version
  attr_accessor :download_url
  attr_accessor :install_url
  attr_accessor :build_url
  attr_accessor :branch_name
  attr_accessor :issue

  def initialize()
    @platform = 'Android'
  end

  def self.from_hash(hash)
    hash = hash || Hash.new
    info = BuildInfo.new
    info.build_id = hash['id']
    info.version = hash['version']
    info.short_version = hash['short_version']
    info.download_url = hash['download_url']
    info.install_url = hash['install_url']
    info
  end
end

def generate_random_build_number
  Time.now.to_i
end

def appcenter_build_dev_url(build_id:)
  return nil if build_id.nil?
  base_url = ENV['APPCENTER_BASE_URL']
  owner_name = ENV['APPCENTER_OWNER_NAME']
  app_name = ENV['APPCENTER_APP_NAME_DEV']
  File.join(base_url, 'orgs', owner_name, 'apps', app_name, 'releases', "#{build_id}")
end

def appcenter_build_prod_url
  base_url = ENV['APPCENTER_BASE_URL']
  owner_name = ENV['APPCENTER_OWNER_NAME']
  app_name = ENV['APPCENTER_APP_NAME']
  group = ENV['APPCENTER_DESTINATIONS_PROD']
  File.join(base_url, 'orgs', owner_name, 'apps', app_name, 'distribution_groups', group)
end

def formatted_build_info_telegram(build_info:)
  text = ""

  unless build_info.nil?
    text += "<strong>Platform:</strong> #{build_info.platform}\n" unless build_info.platform.nil?
    text += "<strong>Build ID:</strong> #{build_info.build_id}\n" unless build_info.build_id.nil?
    text += "<strong>Version:</strong> #{build_info.version}\n" unless build_info.version.nil?
    text += "<strong>Internal Version:</strong> #{build_info.internal_version}\n" unless build_info.internal_version.nil?
    text += "<strong>Short Version:</strong> #{build_info.short_version}\n" unless build_info.short_version.nil?
    text += "<strong>Branch Name:</strong> #{build_info.branch_name}\n" unless build_info.branch_name.nil?
    text += "<strong>Issue:</strong> <a href=\"#{jira_issue_link(key: build_info.issue)}\">#{build_info.issue}</a>\n" unless build_info.issue.nil?
    url_buttons = { "Install URL": build_info.build_url } unless build_info.build_url.nil?
  end
  { "text" => text, "url_buttons" => url_buttons }
end

def formatted_build_info(build_info:)
  result_string = ""
  result_string += "Platform: #{build_info.platform}\n" unless build_info.platform.nil?
  result_string += "Build ID: #{build_info.build_id}\n" unless build_info.build_id.nil?
  result_string += "Version: #{build_info.version}\n" unless build_info.version.nil?
  result_string += "Internal Version: #{build_info.internal_version}\n" unless build_info.internal_version.nil?
  result_string += "Short Version: #{build_info.short_version}\n" unless build_info.short_version.nil?
  result_string += "Branch Name: #{build_info.branch_name}\n" unless build_info.branch_name.nil?
  result_string += "*Install URL* [URL|#{build_info.build_url}]" unless build_info.build_url.nil?
  result_string.strip
end

def formatted_build_info_jira(build_info:)
  result_string = ""
  result_string += "Platform: #{build_info.platform}\n" unless build_info.platform.nil?
  result_string += "*Build ID:* #{build_info.build_id}\n" unless build_info.build_id.nil?
  result_string += "*Version:* #{build_info.version}\n" unless build_info.version.nil?
  result_string += "*Internal Version:* #{build_info.internal_version}\n" unless build_info.internal_version.nil?
  result_string += "*Short Version:* #{build_info.short_version}\n" unless build_info.short_version.nil?
  result_string += "*Branch Name:* #{build_info.branch_name}\n" unless build_info.branch_name.nil?
  result_string += "*Install URL* [URL|#{build_info.build_url}]" unless build_info.build_url.nil?
  result_string.strip
end

File: go/buildInfo.go
package main

import (
	"encoding/json"
	"fmt"
)

type BuildInfo struct {
	Platform         string `json:"platform,omitempty"`
	Build_id         string `json:"BuildID,omitempty"`
	Version          string `json:"Version,omitempty"`
	Internal_version string `json:"InternalVersion,omitempty"`
	External_version string `json:"ExternalVersion,omitempty"`
	Short_version    string `json:"ShortVersion,omitempty"`
	Download_url     string `json:"download_url,omitempty"`
	Install_url      string `json:"InstallURL,omitempty"`
	Build_url        string `json:"build_url,omitempty"`
	Branch_name      string `json:"BranchName,omitempty"`
	Issue            int    `json:"Issue,string,omitempty"`
	TelegramChatId   string `json:"TelegramChatId,string,omitempty"`
	TelegramToken    string `json:"TelegramToken,string,omitempty"`
	ActionId         string `json:"ActionId,string,omitempty"`
	PullNumber       int    `json:"PullNumber,string,omitempty"`
}

func (buildInfo BuildInfo) formatted_build_info_git() string {
	var resultString string

	if buildInfo.Platform != "" {
		resultString += "Platform: " + buildInfo.Platform + "\n"
	}
	if buildInfo.Build_id != "" {
		resultString += "Build ID: " + buildInfo.Build_id + "\n"
	}
	if buildInfo.Version != "" {
		resultString += "Version: " + buildInfo.Version + "\n"
	}
	if buildInfo.Internal_version != "" {
		resultString += "Internal Version: " + buildInfo.Internal_version + "\n"
	}
	if buildInfo.Short_version != "" {
		resultString += "Short Version: " + buildInfo.Short_version + "\n"
	}
	if buildInfo.Branch_name != "" {
		resultString += "Branch Name: " + buildInfo.Branch_name + "\n"
	}
	if buildInfo.Build_url != "" {
		resultString += "Install URL: " + buildInfo.Build_url + "\n"
	}
	return resultString
}

func (buildInfo BuildInfo) build_failed_info(pullNumber string) string {
	var resultString string
	resultString += "🆘<strong>Build failed</strong>🆘" + "\n"
	if buildInfo.Branch_name != "" {
		resultString += "<strong>Branch name: </strong>" + buildInfo.Branch_name + "\n"
	}
	resultString += "<strong>Build Url: </strong>" + "https://github.com/admiral-team/admiralui-ios/pull/" + pullNumber + "/checks" + "\n"
	return resultString
}

func (buildInfo BuildInfo) telegram_release_message() string {
	var resultString string
	resultString += "Новый релиз! ✨" + "\n"
	resultString += "Platform: iOS" + "\n"
	resultString += "Link: https://github.com/admiral-team/admiralui-ios/releases/tag/" + buildInfo.External_version + "\n"
	return resultString
}

func (buildInfo BuildInfo) formatted_build_info_telegram() string {
	var resultString string

	if buildInfo.Platform != "" {
		resultString += "<strong>Platform: </strong>" + buildInfo.Platform + "\n"
	}
	if buildInfo.Build_id != "" {
		resultString += "<strong>Build ID: </strong>" + buildInfo.Build_id + "\n"
	}
	if buildInfo.Version != "" {
		resultString += "<strong>Version: </strong>" + buildInfo.Version + "\n"
	}
	if buildInfo.Internal_version != "" {
		resultString += "<strong>Internal Version: </strong>" + buildInfo.Internal_version + "\n"
	}
	if buildInfo.Short_version != "" {
		resultString += "<strong>Short Version: </strong>" + buildInfo.Short_version + "\n"
	}
	if buildInfo.Branch_name != "" {
		resultString += "<strong>Branch Name: </strong>" + buildInfo.Branch_name + "\n"
	}
	if buildInfo.Build_url != "" {
		resultString += "<strong>Install URL: </strong>" + buildInfo.Build_url + "\n"
	}
	return resultString
}

func configureBuildInfo(byString string) BuildInfo {
	buildInfo := BuildInfo{}
	err := json.Unmarshal([]byte(byString), &buildInfo)

	if err != nil {
		fmt.Println("Cannot Parse JSON From Command Line....", err)
	}
	return buildInfo
}
